File: golang/lib/execution/kurtosis_module_executor.go
/*
 *    Copyright 2021 Kurtosis Technologies Inc.
 *
 *    Licensed under the Apache License, Version 2.0 (the "License");
 *    you may not use this file except in compliance with the License.
 *    You may obtain a copy of the License at
 *
 *        http://www.apache.org/licenses/LICENSE-2.0
 *
 *    Unless required by applicable law or agreed to in writing, software
 *    distributed under the License is distributed on an "AS IS" BASIS,
 *    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *    See the License for the specific language governing permissions and
 *    limitations under the License.
 *
 */

package execution

import (
	"github.com/kurtosis-tech/kurtosis-core-api-lib/api/golang/kurtosis_core_rpc_api_bindings"
	"github.com/kurtosis-tech/kurtosis-core-api-lib/api/golang/lib/enclaves"
	"github.com/kurtosis-tech/kurtosis-core-api-lib/api/golang/module_launch_api"
	grpc_server "github.com/kurtosis-tech/minimal-grpc-server/golang/server"
	"github.com/kurtosis-tech/stacktrace"
	"google.golang.org/grpc"
	"time"
)

const (
	grpcServerStopGracePeriod = 5 * time.Second
)

// Docs available at https://docs.kurtosistech.com/kurtosis-module-api-lib/lib-documentation
type KurtosisModuleExecutor struct {
	configurator KurtosisModuleConfigurator
}

func NewKurtosisModuleExecutor(configurator KurtosisModuleConfigurator) *KurtosisModuleExecutor {
	return &KurtosisModuleExecutor{configurator: configurator}
}

// Docs available at https://docs.kurtosistech.com/kurtosis-module-api-lib/lib-documentation
func (executor KurtosisModuleExecutor) Run() error {
	args, err := module_launch_api.GetArgsFromEnv()
	if err != nil {
		return stacktrace.Propagate(err, "An error occurred getting the module container args from the environment")
	}
	apiContainerSocket := args.ApiContainerSocket
	serializedCustomParams := args.SerializedCustomParams
	enclaveId := enclaves.EnclaveID(args.EnclaveID)
	listenPortNum := args.ListenPortNum
	enclaveDataDirMountpoint := args.EnclaveDataDirMountpoint;

	module, err := executor.configurator.ParseParamsAndCreateExecutableModule(serializedCustomParams)
	if err != nil {
		return stacktrace.Propagate(err, "An error occurred creating the module using serialized custom params '%v'", serializedCustomParams)
	}

	// TODO SECURITY: Use HTTPS to verify we're hitting the correct API container
	conn, err := grpc.Dial(apiContainerSocket, grpc.WithInsecure())
	if err != nil {
		return stacktrace.Propagate(err, "An error occurred dialling the API container at '%v'", apiContainerSocket)
	}

	apiClient := kurtosis_core_rpc_api_bindings.NewApiContainerServiceClient(conn)
	enclaveCtx := enclaves.NewEnclaveContext(
		apiClient,
		enclaveId,
		enclaveDataDirMountpoint,
	)

	serviceImpl := newExecutableModuleServiceImpl(module, enclaveCtx)
	serviceImplRegistrationFunc := func(grpcServer *grpc.Server) {
		kurtosis_core_rpc_api_bindings.RegisterExecutableModuleServiceServer(grpcServer, serviceImpl)
	}

	grpcServer := grpc_server.NewMinimalGRPCServer(
		listenPortNum,
		grpcServerStopGracePeriod,
		[]func(desc *grpc.Server){
			serviceImplRegistrationFunc,
		},
	)
	if err := grpcServer.RunUntilInterrupted(); err != nil {
		return stacktrace.Propagate(err, "An error occurred running the module GRPC server")
	}

	return nil
}

File: typescript/src/lib/execution/kurtosis_module_executor.ts
import { EnclaveContext, getArgsFromEnv } from "kurtosis-core-api-lib";
import { err, ok, Result } from "neverthrow";
import { ExecutableKurtosisModule } from "../kurtosis_modules/executable_kurtosis_modules";
import { KurtosisModuleConfigurator } from "./kurtosis_module_configurator";
import * as grpc from "@grpc/grpc-js";
import { MinimalGRPCServer, TypedServerOverride } from "minimal-grpc-server";
import { ExecutableModuleServiceImpl } from "./executable_module_service_impl";
import { ExecutableModuleServiceService } from "kurtosis-core-api-lib/build/kurtosis_core_rpc_api_bindings/executable_module_service_grpc_pb";

const GRPC_SERVER_STOP_GRACE_PERIOD_SECONDS: number = 5;

// Docs available at https://docs.kurtosistech.com/kurtosis-module-api-lib/lib-documentation
export class KurtosisModuleExecutor {
    constructor(
        private readonly configurator: KurtosisModuleConfigurator,
    ) {}

    // Docs available at https://docs.kurtosistech.com/kurtosis-module-api-lib/lib-documentation
    public async run(): Promise<Result<null, Error>> {
        const getArgsResult = getArgsFromEnv();
        if (getArgsResult.isErr()) {
            return err(new Error(`An error occurred getting the module container args from the environment`));
        }
        const args = getArgsResult.value;
        const apiContainerSocket = args.apiContainerSocket;
        const serializedCustomParams = args.serializedCustomParams;
        const enclaveId = args.enclaveId;
        const listenPortNum = args.listenPortNum;
        const enclaveDataDirMountpoint = args.enclaveDataDirMountpoint;

        const apiContainerSocketFragments: string[] = apiContainerSocket.split(":");
        const ipAddr: string = apiContainerSocketFragments[0]
        const grpcPortNumStr: string = apiContainerSocketFragments[1]
        const grpcPortNum: number = parseInt(grpcPortNumStr)

        const createModuleResult: Result<ExecutableKurtosisModule, Error> = this.configurator.parseParamsAndCreateExecutableModule(serializedCustomParams);
        if (createModuleResult.isErr()) {
            return err(createModuleResult.error);
        }
        const module: ExecutableKurtosisModule = createModuleResult.value;

        // TODO SECURITY: Use HTTPS to verify we're hitting the correct API container
        const createEnclaveCtxResult = await EnclaveContext.newGrpcNodeEnclaveContext(
            ipAddr,
            grpcPortNum,
            enclaveId,
            enclaveDataDirMountpoint,
        );
        if (createEnclaveCtxResult.isErr()) {
            return err(createEnclaveCtxResult.error);
        }
        const enclaveCtx = createEnclaveCtxResult.value;

        const serviceImpl: ExecutableModuleServiceImpl = new ExecutableModuleServiceImpl(
            module,
            enclaveCtx
        );
        const serviceImplRegistrationFunc: { (server: TypedServerOverride): void; }[] = [
            (server: TypedServerOverride) => {
                server.addTypedService(ExecutableModuleServiceService, serviceImpl);
            }
        ];

        const grpcServer: MinimalGRPCServer = new MinimalGRPCServer(
            listenPortNum,
            GRPC_SERVER_STOP_GRACE_PERIOD_SECONDS,
            serviceImplRegistrationFunc
        );
        const runResult: Result<null, Error> = await grpcServer.runUntilInterrupted();
        if (runResult.isErr()) {
            return err(runResult.error);
        }

        return ok(null);
    }
}
